File: sortIntArray.java
//  Timing of Arrays.sort vs Quick vs Shell
//
//  The system sort provided by Java is relatively fast.  Compare this to the
//  system sort provided by Golang, namely sort.Ints, which is disappointingly
//  slow (see sortIntSlice.go).  Shell and Quick each completed sorting in under
//  twice the time of their counterparts in Golang for the same size array, thus
//  there is merit to the claim that Arrays.sort is faster than sort.Ints for
//  sorting an array of integers initially in descending order.
//
//  Sample output:
// 
//      $ go run sortIntSlice.go                $ java sortIntArray.java
//      sort.Ints: 64.323ms                     Arrays.sort:  17 msec
//          Quick: 23.512ms                      Quick Sort:  39 msec
//          Shell: 31.044ms                      Shell Sort:  48 msec

import java.util.*;
import java.lang.*;

class sortIntArray
{
	public static void QuickSort(int a[], int lo, int hi) {
		if (lo >= hi - 1)
			return;

		int pivot = lo + ((hi - lo + 1)>>1);	// ⌈mid-index⌉
		int i = lo, j = hi - 1;
		while (i < j) {
			int temp;

			while (a[i] < a[pivot])
				i++;
			while (a[j] > a[pivot])
				j--;
			if (i < j) {
				temp = a[j];
				a[j] = a[i];
				a[i] = temp;
				i++;
				j--;
			}
		}
		
		QuickSort(a, lo, i - 1);
		QuickSort(a, i + 1, hi);
	}

	public static void ShellSort(int a[]) {
		for (int gap = a.length>>1; gap > 0; gap>>=1) {
			for (int i = gap; i < a.length; i++) {
				int j, temp;

				j = i;
	                        temp = a[i];
	                        for (; j >= gap && a[j-gap] > temp; j -= gap) {
	                                a[j] = a[j-gap];
	                        }
	                        a[j] = temp;
	                }
		}
	}

	public static boolean VerifyAscendingOrder(int a[]) {
		for(int i = 0; i < a.length - 1; i++) {
			if (a[i] > a[i + 1]) {
				System.out.println("failed");
				return (false);
			}
		}
		return (true);
	}

	public static void DescendingOrder(int a[]) {
		for (int i = 0, j = a.length; i < a.length; i++, j--)
			a[i] = j;
	}

	public static void main(String args[]) {
		long bgn, end;
		int[] a = new int [1000000];

		DescendingOrder(a);
		bgn = System.currentTimeMillis(); 
		Arrays.sort(a); 
		end = System.currentTimeMillis(); 
		if (VerifyAscendingOrder(a))
			System.out.println("Arrays.sort:  " + (end - bgn) + " msec");

		DescendingOrder(a);
		bgn = System.currentTimeMillis(); 
		QuickSort(a, 0, a.length);
		end = System.currentTimeMillis(); 
		if (VerifyAscendingOrder(a))
			System.out.println(" Quick Sort:  " + (end - bgn) + " msec");

		DescendingOrder(a);
		bgn = System.currentTimeMillis(); 
		ShellSort(a);
		end = System.currentTimeMillis(); 
		if (VerifyAscendingOrder(a))
			System.out.println(" Shell Sort:  " + (end - bgn) + " msec");

	}
}

File: sortIntArray.cpp
//  Timing of std::array.sort vs Quick vs Shell
//
//  Sorting routines in C++ are expectedly quick because of the absence of portable
//  runtime environments such as RJE for Java and runtime for Golang.  Therefore, it
//  is an apples and oranges comparison.  Nevertheless, the system sort can be tested
//  against other common sorts such as Quick sort and Shell sort to get a rough idea
//  of effiency.  Although comparisons are unfair, array sizes and descending order
//  are identical with sortIntSlice.go and sortIntArray.java for consistency.
//
//  Sample output:
//
//	$ sortArrayInt
//	 std::sort:  0.0609172 msec
//	quick_sort:  0.0393171 msec
//	shell_sort:  0.0734051 msec

#include <iostream>
// Array template
#include <array>
#include <algorithm>
// High resolution timer
#include <ctime>
#include <ratio>
#include <chrono>

using namespace std;

void
quick_sort(int *bgn, int *end)
{
	if (bgn >= end - 1)
		return;

	int *i = bgn;
	int *pivot = bgn + ((end - bgn)>>1);
	int *j = end;
	
	while (i < j) {
		while (*i < *pivot)
			++i;
		while (*j > *pivot) 
			--j;
		if (i < j) {
			int temp = *j;
			*j = *i;
			*i = temp;
			i++;
			j--;
		}
	}

	quick_sort(bgn, i);
	quick_sort(i + 1, end);
}

void
shell_sort(int *bgn, int *end)
{
	for (int *gap = bgn + ((end - bgn)>>1); gap > bgn; gap = bgn + ((gap - bgn)>>1)) {
		for (int *ip = gap; ip < end; ip++) {
			int *jp = ip, temp = *ip;

			for (int *kp = bgn + (jp - gap);
			    jp >= gap && *kp > temp;
			    jp = kp, kp = bgn + (kp - gap)) {
				*jp = *kp;
			}
			*jp = temp;
		}
	}
}

bool
verify_sort(const auto a, string msg)
{
	cout << msg;
	for(int i = 0; i < a.size() - 1; ++i) {
		if (a[i] > a[i + 1]) {
			cout << "Failed\n";
			return (false);
		}
	}
	return (true);
}

void
descending_order(auto &a)
{
	for (int i = 0, j = a.size(); i < a.size(); i++, j--) 
		a[i] = j;
}

void
hi_res_timer(void (*f)(int *, int *), auto &a, string msg)
{
	using namespace std::chrono;

	high_resolution_clock::time_point t1, t2;
	duration<double> time_span;

	descending_order(a);
	t1 = high_resolution_clock::now();
	f(a.begin(), a.end());
	t2 = high_resolution_clock::now();
	time_span = duration_cast<duration<double>>(t2 - t1);
	if (verify_sort(a, msg))
		cout << time_span.count() << " msec\n";
}

int
main()
{
	array <int, 1000000> a;

	hi_res_timer(sort, a, " std::sort:  ");
	hi_res_timer(quick_sort, a, "quick_sort:  ");
	hi_res_timer(shell_sort, a, "shell_sort:  ");
}
